File: XNTM/genlib.py
import json
import copy
import itertools
import sys
import datetime
from pathlib import Path

Mixer = ['6h','6v','4']
Module = ['6h','6v','4','r1','r2','r3','r4','r5']
GridSize = 7

def Msize(M):
    Malt = ["6h","6v","4"]
    size = [[2,3],[3,2],[2,2]]
    
    idx = -1
    for i in range(len(Malt)):
        if Malt[i] == M:
            idx = i
    if idx == -1:
        if M[-1:]=="r":
            return int(M[:-1])
        else :
            print("unexpected input!! in Msize()",file=sys.stderr)
    else : return size[idx]

def numCells(M):
    size = Msize(M)
    ret = 1
    for length in size:
        ret *= length
    return ret 

def gengrid(M):
    grid = [[-1 for h in range(GridSize)]for v in range(GridSize)]
    ParentSize = Msize(M)
    hs,vs = ParentSize
    
    ref_x,ref_y = 2,2 
    for i in range(hs):
        for j in range(vs):
            grid[i+ref_y][j+ref_x] = 0
    return grid

def rootAllCell(M):
    ret = []
    grid = gengrid(M)
    for i in range(GridSize):
        for j in range(GridSize):
            if grid[i][j] == 0:
                ret.append([i,j])
    return ret

### for debugging and checking the generated pattern
def showGrid(grid):
    for y in range(GridSize):
        for x in range(GridSize):
            print("%2d"%(grid[y][x]),end=" ")
        print("\n")

   
def getProvRatio():
    prov_ratio = {}
    ### generate ratio of droplets providing to parent mixer
    for pm in range(len(Mixer)):
        PM = Mixer[pm]
        PMsize = int(PM[0])
        prov_ratio[PM] = []
        q = []
        for i in range(1,6):
            q.append([i])
        while q:
            e = q.pop()
            for i in range(1,6):
                Sum = sum(e)
                nSum = i+Sum
                if nSum > PMsize:
                    continue
                elif nSum == PMsize:
                    ne = copy.deepcopy(e)
                    ne.append(i)
                    insert = sorted(ne,reverse=True)
                    if insert not in prov_ratio[PM]:
                        prov_ratio[PM].append(insert)
                else :
                    ne = copy.deepcopy(e)
                    ne.append(i)
                    q.append(ne)
        #print(prov_ratio[PM],end="\n"*2)
    return prov_ratio

def CanLeftDropPlace(placement,condition):
    search_cell = copy.deepcopy(placement["overlapping_cell"])
    if "flushing_cell" in placement :
        search_cell += placement["flushing_cell"] 
    empty_cell = condition[0][0]
    for cell in search_cell:
        if cell not in empty_cell:
            return False
    return True
    
def LeftDropPlace(placement,condition):
    cells = placement["overlapping_cell"]
    new_condition = copy.deepcopy(condition)

    for cell in cells:
        if cell not in new_condition[0][0]:
            print("error : LeftDropPlace()セルが空いてない!!")
            return -1
        else :
            new_condition[0][0].remove(cell)
            new_condition[0][1].append(cell)
    new_condition[1].append(cells)
    return new_condition
 
def genTemplate(pattern):
    #print(pattern)
    template = {}
    for pm in range(len(Mixer)):
        PM = Mixer[pm]
        template[PM] = {}
        prov_ratio = getProvRatio()
        for ratio in prov_ratio[PM]:
            buffer = [[] for i in range(len(ratio)+1)]
            start = [[rootAllCell(PM),[]],[]]
            buffer[0].append(start)
            
            if numCells(PM) in ratio:
                return
            for idx,num in enumerate(ratio):
                for condition in buffer[idx]:
                    drop = "r"+str(num)
                    for placement in pattern[PM][drop][str(num)]:
                        if CanLeftDropPlace(placement,condition):
                            next_condition = LeftDropPlace(placement,condition)
                            #print(condition,placement,next_condition,sep='\n')
                            buffer[idx+1].append(next_condition)
                #print("\n")
            
            for data in buffer[len(ratio)]:
                if str(ratio) not in template[PM]:
                    template[PM][str(ratio)]=[]
                template[PM][str(ratio)].append(data[1])
            for i in buffer[len(ratio)][1]:
                if not i:
                    print("なんでやねん")
                    return -1
    return template

def showColoredTemplate(prov_ratio,template):
    for pm in range(len(Mixer)):
            PM = Mixer[pm]
            for ratio in prov_ratio[PM]:
                cnt = 0
                for temp in template[PM][str(ratio)]:
                    cnt +=1
                    grid = gengrid(PM)
                    for idx,group in enumerate(temp):
                        for cell in group :
                            color = idx+1
                            y,x = cell
                            grid[y][x] = color
                    print(PM,ratio,sep='\t')
                    showGrid(grid)
                print(PM,ratio,cnt,sep='\t')

def search_placement(pattern,PM,module,drop):
    cnt_drop = len(drop)
    return pattern[PM][module][str(cnt_drop)]
    
def getModule(module):
    FirstLetter = ['6','4','r']
    for i in range(len(FirstLetter)):
        if FirstLetter[i] == module[0]:
            return FirstLetter[i]
    return -1

def main():
    start = datetime.datetime.now()
    readfile = open('XNTM/data/placement.json','r')
    pattern = json.load(readfile)
    prov_ratio = getProvRatio()
    ### generate the combination of placements of dropret left by children-mixers
    template = genTemplate(pattern)
    ### generate library of placement of children-mixers
    lib = {}
    for pm in range(len(Mixer)):
        PM = Mixer[pm]
        lib[PM] = {}
        for ratio in prov_ratio[PM]:
            lib[PM][str(ratio)] = {}
            RatioOrderedCombos = template[PM][str(ratio)]
            for RatioOrderedCombo in RatioOrderedCombos:
                PermuCombo = list(itertools.permutations(RatioOrderedCombo,len(RatioOrderedCombo)))
                for combo in PermuCombo:
                #for combo in RatioOrderedCombo:
                    buffer = [[] for i in range(len(ratio)+1)]
                    ### [[[配置した6ミキサー],[配置した4ミキサー],[配置した試薬液滴]],[試薬液滴を残せるセル],[レイヤーの確認の為のgrid]]
                    buffer[0].append({"Module":{"6":[],"4":[],"r":[]},"CannotPlace":[],"Layer":gengrid(PM)})
                    for idx,DropsLeftByAModule in enumerate(combo):
                        for buff in buffer[idx]:
                        ### DropsLeftByAModuleをFlush後に残すモジュールをbuffer[idx]に追加し，buffer[idx+1]に書き込み
                            for module in range(len(Module)) :
                                MODULE = Module[module]
                                if str(len(DropsLeftByAModule)) in pattern[PM][MODULE]:
                                    for candidate in pattern[PM][MODULE][str(len(DropsLeftByAModule))]:
                                        if candidate["overlapping_cell"] == DropsLeftByAModule:
                                            ### check whether the module overlapps droplets left by mixers that placed formerly
                                            passing = True
                                            check_cells = candidate["flushing_cell"]
                                            for check_cell in check_cells:
                                                if check_cell in buff["CannotPlace"]:
                                                    passing = False
                                            if passing :
                                                new_buff = copy.deepcopy(buff)
                                                layer = 1
                                                layer_search_cells = candidate["flushing_cell"]+candidate["overlapping_cell"]
                                                for layer_search_cell in layer_search_cells:
                                                    y,x = layer_search_cell
                                                    layer = max(layer,buff["Layer"][y][x]+1)
                                                for y,x in layer_search_cells:
                                                    new_buff["Layer"][y][x] = layer
                                                for y,x in candidate["overlapping_cell"]:
                                                    new_buff["CannotPlace"].append([y,x])
                                                AddedModule = copy.deepcopy(candidate)
                                                AddedModule["layer"] = layer
                                                if getModule(MODULE) == '6':
                                                    AddedModule["orientation"] = MODULE[-1:]
                                                new_buff["Module"][getModule(MODULE)].append(AddedModule)
                                                buffer[idx+1].append(new_buff)
                    if buffer[len(ratio)]:
                        for data in buffer[len(ratio)]:
                            modulecnt = [0,0,0]
                            ModuleFirstC = ["6","4","r"]
                            for idx,fc in enumerate(ModuleFirstC):
                                modulecnt[idx] += len(data["Module"][ModuleFirstC[idx]])
                            del data["Layer"]
                            del data["CannotPlace"]
                            if str(modulecnt) not in lib[PM][str(ratio)]:
                                lib[PM][str(ratio)][str(modulecnt)] = []
                            lib[PM][str(ratio)][str(modulecnt)].append(data)
                    else :
                        print("ホンマ，ありえへんてぇ~!! in genlib.py",file=sys.stderr)
                    

    ### output lib to json file
    with open("XNTM/data/data/lib.json","w") as f:
        json.dump(lib,f,indent=4)  

    end = datetime.datetime.now()
    diff = end-start
    print("実行時間 = end - start = {} - {} = {}".format(end,start,diff))

if __name__ == '__main__':
    main()

File: XNTM/genlib.cpp
#include "json.hpp"
#include <algorithm>
#include <fstream>
#include <ios>
#include <iostream>
#include <iterator>
#include <map>
#include <sstream>
#include <string.h>
#include <string>
#include <typeinfo>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
#include <chrono>

using json = nlohmann::json;
using namespace std;

vector<string> Mixer = {"6h", "6v", "4"};
vector<string> Module = {"6h", "6v", "4", "r1", "r2", "r3", "r4", "r5"};
const int GridSize = 7;

pair<int, int> Msize(string Mname) {
  vector<pair<int, int>> size = {{2, 3}, {3, 2}, {2, 2}};
  int i = 0;
  for (auto itr = Mixer.begin(); itr != Mixer.end(); itr++) {
    if (Mname == *itr) {
      return size[i];
    }
  }
  if (Mname[0] == 'r') {
    string digi = Mname.substr(1, Mname.size() - 1);
    int v = stoi(digi);
    return make_pair(v, 1);
  } else
    cerr << "error in Msize()" << endl;
  return make_pair(0, 0);
}

int MCellNum(string Mname) {
  pair<int, int> S = Msize(Mname);
  int v = 1;
  v *= S.first;
  v *= S.second;
  return v;
}

vector<vector<int>> gengrid(string M) {
  vector<vector<int>> grid(GridSize, vector<int>(GridSize, -1));
  pair<int, int> start = make_pair(2, 2);
  pair<int, int> size = Msize(M);
  for (int i = 0; i < size.first; i++) {
    for (int j = 0; j < size.second; j++) {
      int ni = i + start.first;
      int nj = j + start.second;
      grid[ni][nj] = 0;
    }
  }
  return grid;
}

void showgrid(vector<vector<int>> grid) {
  for (int i = 0; i < GridSize; i++) {
    for (int j = 0; j < GridSize; j++) {
      cout << setw(2) << grid[i][j];
    }
    cout << endl;
  }
}

string getModule(string ModuleName) {
  vector<string> FirstChar = {"6", "4", "r"};
  for (int i = 0; i < 3; i++) {
    if (FirstChar[i] == ModuleName.substr(0, 1))
      return FirstChar[i];
  }
  return "0";
}

bool HasKey(string key, json object) {
  auto itr_check = object.find(key);
  if (itr_check != object.end())
    return true;
  else
    return false;
}

bool SameArray(json array, json cmp_array) {
  if (array.size() == cmp_array.size()) {
    for (int i = 0; i < array.size(); i++) {
      if (array[i] != cmp_array[i])
        return false;
    }
    return true;
  }
  return false;
}

void PrintArray(json array) {
  cout << "[";
  for (int i = 0; i < array.size(); i++) {
    cout << array[i] << ",";
  }
  cout << "]" << endl;
}

bool ArrayContainsValue(json value, json Array) {
  for (auto v : Array) {
    if (v == value) {
      return true;
    }
  }
  return false;
}

int main() {
  auto start = chrono::system_clock::now(); 

  ifstream reading1("XNTM/data/PlacementChild.json", ios::in);
  json placementChild;
  reading1 >> placementChild;

  ifstream reading2("XNTM/data/prov_ratio.json", ios::in);
  json prov_ratio;
  reading2 >> prov_ratio;

  ifstream reading3("XNTM/data/PlacementProvDrops.json", ios::in);
  json placementProvDrops;
  reading3 >> placementProvDrops;

  json lib;
  for (string PM : Mixer) {
    lib[PM] = json::object();
    auto now = chrono::system_clock::now(); 
    auto diff = chrono::duration_cast<chrono::hours>(now-start).count(); 
    cout<<"PM:"<<PM<<"の処理に取り掛かる.ここまでの経過時間:"<<diff<<"分"<<endl; 
    for (auto ratio : prov_ratio[PM]) {
      now = chrono::system_clock::now(); 
      diff = chrono::duration_cast<chrono::hours>(now-start).count();
      cout<<"PM:"<<PM<<",ratio:"<<ratio<<"の処理開始.ここまでの経過時間:"<<diff<<"分"<<endl; 
      unordered_map<string,unordered_set<json> > data_set;
      string s_ratio = ratio.dump(); 
      lib[PM][s_ratio] = json::object();

      for (auto Combo : placementProvDrops[PM][s_ratio]) {
        do {
          json buffer;
          for (int i = 0; i < Combo.size() + 1; i++)
            buffer[to_string(i)] = json::array();
          json initial_value = {{"Module",
                                 {{"6", json::array()},
                                  {"4", json::array()},
                                  {"r", json::array()}}},
                                {"CannotPlace", json::array()},
                                {"Layer", gengrid(PM)}};
          buffer[to_string(0)].push_back(initial_value);
          for (int idx = 0; idx < Combo.size(); idx++) {
            json DropsLeftByAModule = Combo[idx];
            for (auto buff : buffer[to_string(idx)]) {
              for (string MODULE : Module) {
                string DropLeftNum = to_string(DropsLeftByAModule.size());
                if (placementChild[PM][MODULE].contains(DropLeftNum)) {
                  for (auto candidate :
                       placementChild[PM][MODULE][DropLeftNum]) {
                    if (SameArray(candidate["overlapping_cell"],
                                  DropsLeftByAModule)) {
                      bool passing = true;
                      json check_cells = candidate["flushing_cell"];
                      for (auto check_cell : check_cells) {
                        if (ArrayContainsValue(check_cell,
                                               buff["CannotPlace"])) {
                          passing = false;
                        }
                      }
                      if (passing) {
                        json new_buff = buff;
                        int layer = 1;
                        json layer_search_cells = candidate["flushing_cell"];
                        for (auto item : candidate["overlapping_cell"])
                          layer_search_cells.push_back(item);
                        for (auto layer_search_cell : layer_search_cells) {
                          int y = layer_search_cell[0];
                          int x = layer_search_cell[1];
                          layer = max(layer, int(buff["Layer"][y][x]) + 1);
                        }
                        for (auto layer_search_cell : layer_search_cells) {
                          int y = layer_search_cell[0];
                          int x = layer_search_cell[1];
                          new_buff["Layer"][y][x] = layer;
                        }
                        for (auto item : candidate["overlapping_cell"]) {
                          new_buff["CannotPlace"].push_back(item);
                        }
                        json AddedModule = candidate;
                        AddedModule["layer"] = layer;
                        if (getModule(MODULE) == "6") {
                          AddedModule["orientation"] = MODULE.substr(1, 1);
                        }
                        string ModuleKind = MODULE.substr(0, 1);
                        new_buff["Module"][ModuleKind]
                            .get_ref<json::array_t &>()
                            .push_back(AddedModule);
                        buffer[to_string(idx + 1)] += new_buff;
                      }
                    }
                  }
                }
              }
            }
          }
          if (!buffer[to_string(Combo.size())].empty()) {
            for (auto data : buffer[to_string(Combo.size())]) {
              vector<int> modulecnt(3, 0);
              vector<string> leftnumPerModule(3, "");
              vector<string> ModuleFirstC = {"6", "4", "r"};
              for (int i = 0; i < ModuleFirstC.size(); i++) {
                modulecnt[i] += int(data["Module"][ModuleFirstC[i]].size());
                // 各モジュールごとにソート(値は，prov_cellの先頭のセル)
                vector<pair<int, json>> v(MCellNum(PM),
                                          make_pair(-1, json::object()));
                for (auto obj : data["Module"][ModuleFirstC[i]]) {
                  json cmp = obj["overlapping_cell"][0];
                  int left_num = obj["overlapping_cell"].size();

                  int y = cmp[0], x = cmp[1];
                  y -= 2;
                  x -= 2;
                  if (getModule(PM) == "6") {
                    if (PM.substr(1, 1) == "h") {
                      v[3 * y + x] = make_pair(left_num, obj);
                    } else {
                      v[2 * y + x] = make_pair(left_num, obj);
                    }
                  } else {
                    v[2 * y + x] = make_pair(left_num, obj);
                  }
                  leftnumPerModule[i] += to_string(left_num);
                }
                json new_array = json::array();
                for (int j = 0; j < MCellNum(PM); j++) {
                  if (v[j].first > 0) {
                    new_array.get_ref<json::array_t &>().push_back(v[j].second);
                  }
                }
                data["Module"][ModuleFirstC[i]] = new_array;
                sort(leftnumPerModule[i].begin(), leftnumPerModule[i].end(),
                     [](char c1, char c2) { return c1 > c2; });
              }
              data.erase(data.find("Layer"));
              data.erase(data.find("CannotPlace"));
              // libに詰め込むときのキー値のひとつとなる各モジュールの数を表す文字列の生成
              string key = "[";
              for (int i = 0; i < ModuleFirstC.size(); i++) {
                key += "[";
                key += to_string(modulecnt[i]);
                key += ",[";
                key += leftnumPerModule[i];
                key += "]],";
              }
              key = key.substr(0, key.size() - 1) + "]";
              if (!lib[PM][s_ratio].contains(key))
                lib[PM][s_ratio][key] = json::array();
              if (!data_set.contains(key))
                  data_set.emplace(key,unordered_set<json>()); 
              if (!data_set[key].contains(data)) {
                data_set[key].emplace(data);
                lib[PM][s_ratio][key].get_ref<json::array_t &>().push_back(data);
              }
            }
          }
          else{
              cout<<"warning:ユニークな配置法が見つからなかった.Comboの順列あり."<<endl; 
          }
        } while (next_permutation(Combo.begin(), Combo.end()));
      }
      ofstream writing_file;
      string WriteFileName = "XNTM/data/lib"+PM+s_ratio+".json"; 
      now = chrono::system_clock::now(); 
      diff = chrono::duration_cast<chrono::hours>(now-start).count();
      cout<<WriteFileName<<"の書き出し.ここまでの経過時間:"<<diff<<"分"<<endl; 
      writing_file.open(WriteFileName,ios::out); 
      writing_file << lib[PM][s_ratio] <<endl; 
      writing_file.close(); 
    }
  }
      ofstream writing_file; 
      string WriteFileName = "XNTM/data/lib.json"; 
      writing_file.open(WriteFileName,ios::out); 
      writing_file << lib <<endl; 
      writing_file.close();
      auto now = chrono::system_clock::now(); 
      auto diff = chrono::duration_cast<chrono::hours>(now-start).count();
      cout<<"genlib処理終了.ここまでの経過時間:"<<diff<<"分"<<endl;
      return 0;
}
